text/0000-packages-as-optional-namespaces.md

```diff
@@ -0,0 +1,229 @@
+foo::render(baz);
+```
+
+Some reasons for `foo`s owner to consider using namespaces:
+- Avoid name conflicts with third-party authors (since they are reserved)
+- Improve discoverability of official crates
+- As an alternative to feature flags for optional subsystems
+- When different parts of your API might have different compatibility guarantees
+
+When considering this, keep in mind:
+- Does it makes sense for this new crate to be presented in the `foo` facade?
+- How likely is a crate to move into or out of the namespace?
+  - Moving the crate in or out of a namespace is a breaking change though it can be worked around by having the old crate re-export the new crate but that does add extra friction to the process.
+  - There is not currently a mechanism to raise awareness with users that a crate has migrated into or out of a namespace and you might end up leaving users behind.
+- If users import both `foo` and `foo::bar` but `foo` also has a `bar` item in its API that isn't just `foo::bar` re-exported, then rustc will error.
+
+Only the owners of `foo` may _create_ the `foo::bar` crate (and all owners of `foo` are implicitly owners of `foo::bar`). After the `foo::bar` crate is created, additional per-crate publishers may be added who will be able to publish subsequent versions as usual.
+
+# Reference-level explanation
+
+`::` is now considered valid inside crate names on Crates.io. For now, we will restrict crate names to having a single `::` in them, not at the beginning or end of the name, but this can be changed in the future.
+
+When publishing a crate `foo::bar`, if the crate does not exist, the following must be true:
+
+ - `foo` must exist
+ - The user publishing the crate must be an owner of `foo`
+
+For the crate `foo::bar`, all owners of `foo` are always considered owners of `foo::bar`, however additional owners may be added. People removed from ownership of `foo` will also lose access to `foo::bar` unless they were explicitly added as owners to `foo::bar`.
+
+Crates.io displays `foo::bar` crates with the name `foo::bar`, though it may stylistically make the `foo` part link to the `foo` crate.
+
+The [registry index trie](https://doc.rust-lang.org/nightly/cargo/reference/registries.html#index-format) may represent subpackages by placing `foo::bar` as just `foo::bar`.
+
+`rustc` will need some changes. When `--extern foo::bar=crate.rlib` is passed in, `rustc` will include this crate during resolution as if it were a module `bar` living under crate `foo`. If crate `foo` is _also_ in scope, this will not automatically trigger any errors unless `foo::bar` is referenced, `foo` has a module `bar`, and that module is not just a reexport of crate `foo::bar`.
+
+The autogenerated `lib.name` key for such a crate will just be `bar`, the leaf crate name, and the expectation is that to use such crates one _must_ use `--extern foo::bar=bar.rlib` syntax. There may be some better things possible here, perhaps `foo_bar` can be used here.
```

Comment:
I'm assuming the following will be legal
```toml
"foo::bar" = { package = "foo_bar", version = "1.0" }
```
Should we let this remain legal or add a check to disallow this?

Reply:
I'm ambivalent

Reply:
@pitaj brought up a reason to allow this

> > For example, if someone were to add `evil-serde::derive = { package = "serde::derive" }`, would this now be importable as `serde::derive`?
> 
> It would be the opposite:
> 
> ```
> "serde::derive" = { package = "evil-serde::derive", version = 1 }
> ```
> 
> I think it's important to allow this so people can use git dependencies, local paths, and drop-in replacements in place of any crate.

Reply:
Yeah I think git dependencies and local paths are an absolute requirement. Allowing drop-in replacements maybe not so much, but it would be kinda weird to forbid just that unless we have a good reason.

Reply:
Added this to unresolved questions. I'm overall okay with this being banged out as a part of the experimentation process.

Reply:
There are very few crates with `__` in their name. Double underscore could be realistically be reserved in crate names, and then function as a namespace separator that is compatible with all file systems and distro packages.